File: extraccion/geometria.py
"""Los dos sistemas de coordenadas, y el criterio de traslape del 50%.

Un PDF mide en puntos con el origen ABAJO a la izquierda. Una imagen rasterizada
mide en pixeles con el origen ARRIBA a la izquierda. Todo el trabajo de este
paquete consiste en cruzar anotaciones del PDF con texto leido de la imagen, asi
que las dos cosas tienen que hablar el mismo idioma antes de compararse.

Aqui se convierte a pixeles de imagen, y se convierte una sola vez.
"""

import math
from collections import namedtuple

# Rectangulo en pixeles de imagen: origen arriba a la izquierda, `y0` es el borde
# de arriba y `y1` el de abajo. Siempre `y0 <= y1` y `x0 <= x1`.
Rectangulo = namedtuple("Rectangulo", ("x0", "y0", "x1", "y1"))

# Regla de no regresion (`DECISIONES.md`): sin tope, un escaneo grande a 300 DPI
# tardaba entre 5 y 10 minutos por pagina.
LADO_LARGO_MAXIMO_PX = 3500


def escala_de_rasterizado(ancho_puntos, alto_puntos, lado_largo_maximo_px=LADO_LARGO_MAXIMO_PX):
    """La escala que deja el lado largo justo EN el tope, nunca uno por encima.

    El `math.nextafter` no es un adorno y no se puede simplificar a una division
    (`DECISIONES.md`, FASE 0): `3500/792` vale 3500.0000000000005 en coma
    flotante, pypdfium2 redondea el tamano del mapa de bits, y la pagina sale a
    3501 px. Bajar la escala al float inmediatamente anterior la deja en 3500.
    `pruebas/prueba_geometria.py` fija los dos numeros.
    """
    lado_largo_puntos = max(ancho_puntos, alto_puntos)
    return math.nextafter(lado_largo_maximo_px / lado_largo_puntos, 0.0)


def rectangulo_pdf_a_pixeles(rectangulo_pdf, alto_pagina_puntos, escala):
    """Pasa un `/Rect` de anotacion a pixeles de la imagen rasterizada.

    `rectangulo_pdf` llega como `(x1, y1, x2, y2)` en puntos, con `y` creciendo
    hacia ARRIBA. La `y` se invierte restandola del alto de la pagina; la `x`
    solo se escala, porque comparte origen en los dos sistemas.

    Omitir la inversion deja los rectangulos espejados respecto al centro de la
    pagina y el sistema anula campos que estaban buenos. Por eso la conversion
    esta aislada en esta funcion y tiene prueba propia.
    """
    x_izquierda_pdf, y_abajo_pdf, x_derecha_pdf, y_arriba_pdf = (
        float(valor) for valor in rectangulo_pdf
    )
    return Rectangulo(
        x0=min(x_izquierda_pdf, x_derecha_pdf) * escala,
        y0=(alto_pagina_puntos - max(y_abajo_pdf, y_arriba_pdf)) * escala,
        x1=max(x_izquierda_pdf, x_derecha_pdf) * escala,
        y1=(alto_pagina_puntos - min(y_abajo_pdf, y_arriba_pdf)) * escala,
    )


def rectangulo_desde_puntos(puntos):
    """La caja que envuelve una lista de puntos `(x, y)` en pixeles de imagen.

    Es lo que devuelve el OCR para cada linea: cuatro esquinas, no un rectangulo.
    """
    equis = [float(punto[0]) for punto in puntos]
    yes = [float(punto[1]) for punto in puntos]
    return Rectangulo(x0=min(equis), y0=min(yes), x1=max(equis), y1=max(yes))


def traslape_vertical(rectangulo, banda):
    """Los pixeles de alto que los dos rectangulos comparten. Nunca negativo."""
    return max(0.0, min(rectangulo.y1, banda.y1) - max(rectangulo.y0, banda.y0))


def fraccion_de_traslape_vertical(rectangulo, banda):
    """Que parte del ALTO DE LA BANDA cubre el rectangulo, entre 0.0 y 1.0.

    El denominador es la altura de la banda y no la del rectangulo: asi lo fija
    `DECISIONES.md`. Una banda de alto cero devuelve 0.0 en vez de reventar.
    """
    alto_de_la_banda = banda.y1 - banda.y0
    if alto_de_la_banda <= 0.0:
        return 0.0
    return traslape_vertical(rectangulo, banda) / alto_de_la_banda


def pertenece_a_la_banda(rectangulo, banda, fraccion_minima=0.5):
    """Cierto cuando el traslape SUPERA la fraccion pedida.

    La comparacion es estricta (`>`), no `>=`: `DECISIONES.md` dice «supera el
    50%», y con el empate exacto la banda no se anula. `prueba_geometria.py`
    prueba los dos lados del borde y el empate.
    """
    return fraccion_de_traslape_vertical(rectangulo, banda) > fraccion_minima


def se_solapan_en_horizontal(rectangulo, banda):
    """Cierto cuando los dos comparten al menos un pixel de ancho.

    Sirve para no dejar que un tachon de la columna derecha anule un campo de la
    columna izquierda: el formulario tiene dos columnas y una banda no ocupa la
    pagina entera.
    """
    return min(rectangulo.x1, banda.x1) > max(rectangulo.x0, banda.x0)

File: csharp/Fichas/Fichas.Lectura/Geometria.cs
using Fichas.Contratos.Lectura;

namespace Fichas.Lectura;

/// <summary>
/// Los dos sistemas de coordenadas, la escala del rasterizado y el traslape del 50%.
/// </summary>
/// <remarks>
/// Portado de `extraccion/geometria.py`. Un PDF mide en puntos con el origen ABAJO a la
/// izquierda; una imagen rasterizada mide en pixeles con el origen ARRIBA. Todo el
/// trabajo consiste en cruzar anotaciones del PDF con texto leido de la imagen, asi que
/// las dos cosas hablan el mismo idioma antes de compararse.
///
/// <para><b>Aqui se trabaja en FRACCIONES de pagina</b> y no en pixeles, porque es lo que
/// dice <see cref="BandaDeLaPagina"/>: un pixel depende de la escala con la que se
/// rasterizo ese dia, y una fraccion sigue valiendo a cualquier escala.</para>
/// </remarks>
public static class Geometria
{
    /// <summary>
    /// El tope del lado largo al rasterizar. Regla de no regresion: sin tope, un escaneo
    /// a 300 DPI tardaba entre 5 y 10 minutos por pagina.
    /// </summary>
    public const int LadoLargoMaximoPx = 3500;

    /// <summary>
    /// La escala que deja el lado largo justo EN el tope, nunca uno por encima.
    /// </summary>
    /// <remarks>
    /// El <c>BitDecrement</c> no es un adorno y no se puede simplificar a una division
    /// (`DECISIONES.md`, FASE 0): <c>3500/792</c> vale 3500,0000000000005 en coma
    /// flotante, el rasterizador redondea hacia arriba el tamano del mapa de bits, y la
    /// pagina sale a 3501 px. Bajar la escala al <c>double</c> inmediatamente anterior la
    /// deja en 3500. Es el equivalente exacto del <c>math.nextafter(..., 0.0)</c> del
    /// Python, y `PruebaDeLecturaDePdf` fija los dos numeros.
    /// </remarks>
    /// <param name="anchoPuntos">Ancho de la página en puntos PDF.</param>
    /// <param name="altoPuntos">Alto de la página en puntos PDF.</param>
    /// <param name="ladoLargoMaximoPx">El tope en píxeles del lado más largo; las pruebas lo cambian para medir el redondeo.</param>
    /// <returns>El factor por el que se multiplican los puntos para obtener píxeles.</returns>
    public static double EscalaDeRasterizado(double anchoPuntos, double altoPuntos, int ladoLargoMaximoPx = LadoLargoMaximoPx)
        => Math.BitDecrement(ladoLargoMaximoPx / Math.Max(anchoPuntos, altoPuntos));

    /// <summary>
    /// Pasa un <c>/Rect</c> de anotacion —en puntos, con la <c>y</c> creciendo hacia
    /// ARRIBA— a fracciones de la pagina, con la <c>y</c> creciendo hacia abajo.
    /// </summary>
    /// <remarks>
    /// Omitir la inversion de la <c>y</c> deja los rectangulos espejados respecto al
    /// centro de la pagina, y el sistema anula campos que estaban buenos. Por eso la
    /// conversion esta aislada aqui y tiene prueba propia.
    /// </remarks>
    /// <param name="izquierda">Primera coordenada <c>x</c> del <c>/Rect</c>; si viene al revés que <paramref name="derecha"/> se ordenan.</param>
    /// <param name="abajo">Primera coordenada <c>y</c> del <c>/Rect</c>, medida desde el pie de la página.</param>
    /// <param name="derecha">Segunda coordenada <c>x</c> del <c>/Rect</c>.</param>
    /// <param name="arriba">Segunda coordenada <c>y</c> del <c>/Rect</c>.</param>
    /// <param name="anchoPuntos">Ancho de la página en puntos.</param>
    /// <param name="altoPuntos">Alto de la página en puntos.</param>
    /// <returns>La caja en fracciones 0–1 con el origen arriba; con una página de tamaño cero o negativo, la caja nula en vez de dividir por cero.</returns>
    public static BandaDeLaPagina RectanguloPdfAFracciones(
        double izquierda, double abajo, double derecha, double arriba, double anchoPuntos, double altoPuntos)
    {
        if (anchoPuntos <= 0.0 || altoPuntos <= 0.0) return new BandaDeLaPagina(0.0, 0.0, 0.0, 0.0);

        return new BandaDeLaPagina(
            X0: Math.Min(izquierda, derecha) / anchoPuntos,
            Y0: (altoPuntos - Math.Max(abajo, arriba)) / altoPuntos,
            X1: Math.Max(izquierda, derecha) / anchoPuntos,
            Y1: (altoPuntos - Math.Min(abajo, arriba)) / altoPuntos);
    }

    /// <summary>La caja que envuelve una lista de puntos en pixeles, ya en fracciones.</summary>
    /// <param name="puntos">Las esquinas del polígono que devuelve el detector del OCR, en píxeles de la imagen.</param>
    /// <param name="anchoPx">Ancho de la imagen rasterizada.</param>
    /// <param name="altoPx">Alto de la imagen rasterizada.</param>
    /// <returns>El rectángulo mínimo que los contiene, en fracciones; la caja nula si no hay puntos o la imagen no tiene tamaño.</returns>
    public static BandaDeLaPagina BandaDesdePuntos(IEnumerable<(double X, double Y)> puntos, int anchoPx, int altoPx)
    {
        if (anchoPx <= 0 || altoPx <= 0) return new BandaDeLaPagina(0.0, 0.0, 0.0, 0.0);

        double x0 = double.MaxValue, y0 = double.MaxValue, x1 = double.MinValue, y1 = double.MinValue;
        foreach (var (x, y) in puntos)
        {
            if (x < x0) x0 = x;
            if (y < y0) y0 = y;
            if (x > x1) x1 = x;
            if (y > y1) y1 = y;
        }
        if (x0 > x1) return new BandaDeLaPagina(0.0, 0.0, 0.0, 0.0);

        return new BandaDeLaPagina(x0 / anchoPx, y0 / altoPx, x1 / anchoPx, y1 / altoPx);
    }

    /// <summary>Lo que los dos rectangulos comparten de alto. Nunca negativo.</summary>
    /// <param name="rectangulo">Una caja en fracciones de página.</param>
    /// <param name="banda">La otra; el orden no cambia el resultado.</param>
    /// <returns>El alto común en fracciones de página, o 0,0 si no se tocan en vertical.</returns>
    public static double TraslapeVertical(BandaDeLaPagina rectangulo, BandaDeLaPagina banda)
        => Math.Max(0.0, Math.Min(rectangulo.Y1, banda.Y1) - Math.Max(rectangulo.Y0, banda.Y0));

    /// <summary>
    /// Que parte del ALTO DE LA BANDA cubre el rectangulo, entre 0,0 y 1,0.
    /// </summary>
    /// <remarks>
    /// El denominador es la altura de la BANDA y no la del rectangulo: asi lo fija
    /// `DECISIONES.md`. Una banda de alto cero devuelve 0,0 en vez de reventar.
    /// </remarks>
    /// <param name="rectangulo">La caja de una línea del OCR o de una anotación.</param>
    /// <param name="banda">La fila de valor cuyo alto hace de denominador.</param>
    public static double FraccionDeTraslapeVertical(BandaDeLaPagina rectangulo, BandaDeLaPagina banda)
    {
        double altoDeLaBanda = banda.Y1 - banda.Y0;
        return altoDeLaBanda <= 0.0 ? 0.0 : TraslapeVertical(rectangulo, banda) / altoDeLaBanda;
    }

    /// <summary>
    /// Que parte del ALTO DEL RECTANGULO cae dentro de la banda, entre 0,0 y 1,0.
    /// </summary>
    /// <remarks>
    /// Es <see cref="FraccionDeTraslapeVertical"/> con el denominador al reves, y existe
    /// porque las dos preguntas son distintas y las dos hacen falta:
    /// <list type="bullet">
    ///   <item>«¿cubre este rectangulo la banda?» decide si un TACHON la anula, y por eso
    ///   mide contra el alto de la banda: un trazo fino tiene que poder anular una fila.</item>
    ///   <item>«¿cabe esta linea en la banda?» decide si un texto es de esa fila del papel,
    ///   y tiene que medir contra el alto de la LINEA: una caja del OCR que se tragó tres
    ///   renglones cubre el 100% de cualquier banda que cruce, y sin esta pregunta su texto
    ///   se le da al campo de cada una.</item>
    /// </list>
    /// <para>Medido el 2026-09-07 sobre las 20 hojas reales, en las bandas de los tres
    /// campos del caso que se extraen: las lineas que de verdad son de esa fila dan de
    /// <b>0,588 a 1,000</b>; las que son de otra fila dan de <b>0,354 a 0,481</b>.</para>
    /// <para>Un rectangulo de alto cero devuelve 0,0 en vez de reventar.</para>
    /// </remarks>
    /// <param name="rectangulo">La caja de una línea del OCR, cuyo alto hace de denominador.</param>
    /// <param name="banda">La fila de valor que cuelga del ancla.</param>
    public static double FraccionDelRectanguloDentroDeLaBanda(BandaDeLaPagina rectangulo, BandaDeLaPagina banda)
    {
        double altoDelRectangulo = rectangulo.Y1 - rectangulo.Y0;
        return altoDelRectangulo <= 0.0 ? 0.0 : TraslapeVertical(rectangulo, banda) / altoDelRectangulo;
    }

    /// <summary>
    /// Cierto cuando los dos comparten al menos algo de ancho.
    /// </summary>
    /// <remarks>
    /// Sirve para no dejar que un tachon de la columna derecha anule un campo de la
    /// columna izquierda: el formulario tiene dos columnas y una banda no ocupa la
    /// pagina entera.
    /// </remarks>
    /// <param name="rectangulo">Una caja en fracciones de página.</param>
    /// <param name="banda">La otra; tocarse solo en el borde no cuenta como solapar.</param>
    public static bool SeSolapanEnHorizontal(BandaDeLaPagina rectangulo, BandaDeLaPagina banda)
        => Math.Min(rectangulo.X1, banda.X1) > Math.Max(rectangulo.X0, banda.X0);
}
